packages/xrpc/src/util.ts

```diff
@@ -96,40 +106,191 @@ export function normalizeHeaders(headers: Headers): Headers {
+      ) {
+        headers['content-type'] = 'application/json'
+      } else {
+        // symbol, function, bigint
+        throw new XRPCError(
+          ResponseType.InvalidRequest,
+          `Unsupported data type: ${typeof data}`,
+        )
       }
     }
   }
   return headers
 }
 
+export function isBodyInit(value: unknown): value is BodyInit {
+  switch (typeof value) {
+    case 'string':
+      return true
+    case 'object':
+      return (
+        value instanceof ArrayBuffer ||
+        value instanceof Blob ||
+        value instanceof FormData ||
+        value instanceof URLSearchParams ||
+        value instanceof ReadableStream ||
+        ArrayBuffer.isView(value)
+      )
+    default:
+      return false
+  }
+}
+
+export function isIterable(
+  value: unknown,
+): value is Iterable<unknown> | AsyncIterable<unknown> {
+  return (
+    typeof value === 'object' &&
+    value !== null &&
+    (Symbol.iterator in value || Symbol.asyncIterator in value)
+  )
+}
+
 export function encodeMethodCallBody(
   headers: Headers,
-  data?: any,
-): ArrayBuffer | undefined {
-  if (!headers['content-type'] || typeof data === 'undefined') {
+  data?: unknown,
+): BodyInit | undefined {
+  // Silently ignore the body if there is no content-type header.
+  if (!headers['content-type']) {
     return undefined
   }
-  if (data instanceof ArrayBuffer) {
+
+  if (typeof data === 'undefined') {
+    // This error would be returned by the server, but we can catch it earlier
+    // to avoid un-necessary requests. Note that a content-length of 0 does not
+    // necessary mean that the body is "empty" (e.g. an empty txt file).
+    throw new XRPCError(
+      ResponseType.InvalidRequest,
+      `A request body is expected but none was provided`,
+    )
+  }
+
+  if (isBodyInit(data)) {
+    if (data instanceof FormData) {
+      // fetch() will encode FormData payload itself, but it won't override the
+      // content-type header if already present. This would cause the boundary
+      // to be missing from the content-type header, resulting in a 400 error.
+      // Deleting the content-type header here to let fetch() re-create it.
+      delete headers['content-type']
+    }
+
+    // Will be encoded by the fetch API.
     return data
   }
+
+  if (isIterable(data)) {
+    // Note that some environments support using Iterable / AsyncIterable as the
+    // body directly (Node's fetch), but not all (browsers). Let's make sure we
+    // always return a ReadableStream. ReadableStream.from is not available in
+    // all environments. Luckily, it is quite easy to implement ourselves.
+
+    // https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/from_static
+    // return ReadableStream.from(data)
+
+    const getIterator = async function* () {
+      for await (const part of data) yield part
+    }
+
+    let it: AsyncGenerator<unknown, void, undefined>
+    return new ReadableStream<Uint8Array>({
```

Comment:
Could be nice to extract this conversion from iterable to readable stream into a helper.